ons-spark/spark-concepts/shuffling.md

```diff
@@ -0,0 +1,755 @@
+| 19|           1|    8|
++---+------------+-----+
+```
+Comparing this to the previous output, we see that the `partition_id` has not changed. Why? This is because the data did not need to be shuffled between the partitions, as the creation of a column of random numbers is a *narrow transformation*. This calculation can be done independently on each row and so all partitions can be processed in parallel.
+
+Now try sorting the data by `rand1` in ascending order. Sorting will cause a shuffle, as the data needs to be moved between partitions to get it in the correct order. Then create a column for the `partition_id` after the sort, called `partition_id_new`.
+````{tabs}
+```{code-tab} py
+example_1 = (example_1
+             .orderBy("rand1")
+             .withColumn("partition_id_new", F.spark_partition_id()))
+
+example_1.show()
+```
+
+```{code-tab} r R
+
+example_1 <- example_1 %>%
+    sparklyr::sdf_sort("rand1") %>%
+    sparklyr::mutate(partition_id_new = spark_partition_id())
+
+example_1 %>%
+    sparklyr::collect() %>%
+    print()
+
+```
+````
+
+```plaintext
++---+------------+-----+----------------+
+| id|partition_id|rand1|partition_id_new|
++---+------------+-----+----------------+
+| 10|           1|    1|               0|
+| 16|           1|    1|               0|
+|  5|           0|    2|               0|
+|  6|           0|    2|               0|
+|  9|           0|    3|               0|
+|  1|           0|    4|               0|
+|  2|           0|    4|               0|
+|  8|           0|    4|               0|
+| 11|           1|    4|               0|
+| 14|           1|    4|               0|
+|  4|           0|    6|               1|
+| 13|           1|    6|               1|
+| 17|           1|    6|               1|
+|  3|           0|    8|               1|
+| 19|           1|    8|               1|
+|  0|           0|    9|               1|
+| 12|           1|    9|               1|
+| 15|           1|    9|               1|
+| 18|           1|    9|               1|
+|  7|           0|   10|               1|
++---+------------+-----+----------------+
+```
+We can see that some data has moved between the partitions. Specifically, any row where `partition_id` $ \neq$ `partition_id_new`. Note that not all the data has changed partitions; some are already on the correct partition prior to the sorting.
+
+As we specified `spark.sql.shuffle.partitions` to be  `2` in the config, the shuffle has returned two partitions. In practical usage you will almost never only use two partitions, but this principle applies regardless of the size of the DF or number of partitions.
+
+Another way to see when a shuffle has occurred is to check the Spark UI, where a shuffle is referred to as an **Exchange**. This is covered in more detail in the Spark Application and UI article.
+
+Generate a link to the Spark UI:
+````{tabs}
+```{code-tab} py
+spark_ui_url = "spark-%s.%s" % (os.environ["CDSW_ENGINE_ID"], os.environ["CDSW_DOMAIN"])
```

Comment:
we need to make a decision on Spark UI URLs. No answer from Nigel, I could nudge him, but we could also just use the local URL so others could use the notebooks